mixer/adapter/dogstatsd/dogstatsd.go

```diff
@@ -0,0 +1,207 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:generate $GOPATH/src/istio.io/istio/bin/mixer_codegen.sh -f mixer/adapter/dogstatsd/config/config.proto
+package dogstatsd
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	multierror "github.com/hashicorp/go-multierror"
+
+	descriptor "istio.io/api/mixer/v1/config/descriptor"
+	"istio.io/istio/mixer/adapter/dogstatsd/config"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/metric"
+)
+
+type builder struct {
+	adapterConfig *config.Params
+	metricTypes   map[string]*metric.Type
+}
+type handler struct {
+	rate    float64
+	client  *statsd.Client
+	metrics map[string]*config.Params_MetricInfo
+}
+
+///////////////// Configuration-time Methods ///////////////
+
+// adapter.HandlerBuilder#Build
+func (b *builder) Build(_ context.Context, env adapter.Env) (adapter.Handler, error) {
+	ac := b.adapterConfig
+
+	client := &statsd.Client{}
+	err := error(nil)
+	if ac.BufferLength > 0 {
+		client, err = statsd.NewBuffered(ac.Address, int(ac.BufferLength))
+	} else {
+		client, err = statsd.New(ac.Address)
+	}
+	if err != nil {
+		env.Logger().Errorf("Unable to create dogstatsd client: %v", err)
+		return nil, err
+	}
+	client.Namespace = ac.Prefix
+	client.Tags = flattenTags(ac.GlobalTags)
+
+	// Generate a warning for metrics that the Mixer emits but aren't handled by the adapter
+	for mname := range b.metricTypes {
+		if _, found := ac.Metrics[mname]; !found {
+			env.Logger().Warningf("%s is a valid metric but is not configured to be handled byt the datadog adapter", mname)
```

Comment:
byt -> by

This check is already done in Validate, right?

The Mixer contract is that it won't call Build unless Validate succeeds.

Reply:
Validate checks that the adapterConfig metrics exist as a valid metric type while this is looking at the opposite. I was going to add this to validate but 1. didn't think it necessitates failing validation in this case and 2. validate doesn't have access to the env.Logger to generate a warning (unless I've missed something).

Reply:
You're right, the Validate function doesn't have access to an env logger. All validation errors should be returned in the error.

The idea with Validate is that anything wrong in the config entered by the operator should be reported there. This will prevent bad config from entering the system at all. By the time Build runs, the only errors we should be producing are the transient variety around network connectivity or credentials. Things that are strictly related to operator errors should be in Validate.

In general, we want to encourage config to be well-formed and strongly correct in order to eliminate ambiguities and issues with future upgradability. So we want to be aggressive about flagging errors.

Reply:
Stepping back a bit. When `SetMetricsType` gets called on the builder, are the metrics that are  passed in all of the metric templates that the mixer knows about or just the metrics that have rules that map them to the adapter? My first thought was that it was all metrics and it made sense to me to log and continue in that case. 

If `SetMetricsType` is only passing specific metrics to the handler and the handler doesn't have the corresponding config then, yes a validation error makes sense there.